tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslEngine.java

```diff
@@ -46,34 +41,17 @@
 
     /**
      * Instrument the provided {@link SSLEngine}.
-     * This method behaves in a strange way to support java 9 without fooling java 8 consumers. Many libraries use
-     * the existence of the new JDK9 methods to attempt ALPN via reflection, so we cannot return an implementation
-     * with these methods unless the delegate engine provides them. Conversely, regardless of the runtime version
-     * we need to preserve the new ALPN methods from the delegate engine to avoid modifying application behavior.
+     * Note that this assumes up-to-date JVMs, older Java 8 releases did not provide
+     * ALPN API methods {@link #getApplicationProtocol()}, {@link #getHandshakeApplicationProtocol()},
+     * {@link #setHandshakeApplicationProtocolSelector(BiFunction)}, {@link #getHandshakeApplicationProtocolSelector()}.
+     * Using this method on an older jre8 release will not fail outright, but some callers may assume
+     * ALPN is supported by this implementation when calls to the delegate fail. Reflection is not
+     * used here because it increases risk of incorrect behavior using graalvm native-images.
+     *
+     * @see <a href="https://openjdk.java.net/jeps/244">JEP 244</a>
+     * @see <a href="https://bugs.openjdk.java.net/browse/JDK-8230977">JDK-8230977</a>
      */
     static SSLEngine instrument(SSLEngine engine, TlsMetrics metrics, String name) {
```

Comment:
Given ALPN was backported to 8u252 per https://bugs.openjdk.java.net/browse/JDK-8230977 so should we do something where we don't instrument on JDK 1.8 < 8u252, e.g. checking `java.version` system property rather than the reflection checks?

```suggestion
     * Note that this assumes up-to-date JVMs, older Java 8 releases did not provide
     * ALPN API methods {@link #getApplicationProtocol()}, {@link #getHandshakeApplicationProtocol()},
     * {@link #setHandshakeApplicationProtocolSelector(BiFunction)}, {@link #getHandshakeApplicationProtocolSelector()}.
     * Using this method on an older jre8 release will not fail outright, but some callers may assume
     * ALPN is supported by this implementation when calls to the delegate fail. Reflection is not
     * used here because it increases risk of incorrect behavior using graalvm native-images.
     *
     * @see <a href="https://openjdk.java.net/jeps/244">JEP 244</a>
     * @see <a href="https://bugs.openjdk.java.net/browse/JDK-8230977">JDK-8230977</a>
     */
    static SSLEngine instrument(SSLEngine engine, TlsMetrics metrics, String name) {
        if (hasAlpn()) {
            return engine;
        }
```

and

```
    private static final Pattern java8MinorVersionRegex = Pattern.compile("1.8.0_([0-9]+)");

    private static boolean hasAlpn() {
        OptionalInt minorUpdateVersion = parseMinorUpdateVersion(System.getProperty("java.version"));
        if (minorUpdateVersion.isPresent()) {
            // ALPN support exists in JDK 8u252+ https://bugs.openjdk.java.net/browse/JDK-8230977
            return minorUpdateVersion.getAsInt() >= 252;
        }
        return true; // ALPN exists in JDK 9+ https://bugs.openjdk.java.net/browse/JDK-8051498
    }

    private static OptionalInt parseMinorUpdateVersion(String javaVersion) {
        if (javaVersion != null) {
            Matcher matcher = java8MinorVersionRegex.matcher(javaVersion);
            if (matcher.matches()) {
                try {
                    return OptionalInt.of(Integer.parseInt(matcher.group(1)));
                } catch (NumberFormatException e) {
                    return OptionalInt.empty();
                }
            }
        }
        return OptionalInt.empty();
    }
```

Reply:
I considered a conditional here, we've written [similar checks before](https://github.com/apache/httpcomponents-core/pull/215), but I'm not sure it's worthwhile at this point for a few reasons:
* Our use of java 8 is waning
* Java 8 builds which don't support ALPN are considered insecure
* We supported conditionally implementing these methods to work sympathetically with other libraries, however those libraries have been updated to make that unnecessary: https://github.com/undertow-io/undertow/pull/977

Reply:
Happy to add the switch if you prefer that approach though, I don't have a terribly strong opinion.

Reply:
yeah, I think I'm fine if we assume at least 8u252+ and if folks hit `NoSuchMethodError` or other odd behavior the solution is upgrade to supported JDK runtime.